### PR TITLE
[FIR] Introduce FirInitializerTypeMismatch & FirReturnTypeMismatch checkers

### PR DESCRIPTION
**Pull request by vldF**

### CODE REVIEW

File: compiler/frontend/src/org/jetbrains/kotlin/diagnostics/PositioningStrategies.kt [hunk 1]
```diff
@@ -759,6 +759,26 @@ object PositioningStrategies {
     val REFERENCE_BY_QUALIFIED: PositioningStrategy<PsiElement> = FindReferencePositioningStrategy(false)
     val REFERENCED_NAME_BY_QUALIFIED: PositioningStrategy<PsiElement> = FindReferencePositioningStrategy(true)
 
+    val ASSIGNMENT_VALUE: PositioningStrategy<KtProperty> = object : PositioningStrategy<PsiElement>() {
+        override fun mark(element: PsiElement): List<TextRange> {
+            return if (element is KtProperty) {
+                markElement(element.initializer ?: element)
+            } else {
+                markElement(element)
+            }
```
Comment (KvanTTT) [inline compiler/frontend/src/org/jetbrains/kotlin/diagnostics/PositioningStrategies.kt:768]: Simplify?

```suggestion
            return markElement(if (element is KtProperty) element.initializer ?: element else element)
```

File: compiler/frontend/src/org/jetbrains/kotlin/diagnostics/PositioningStrategies.kt [hunk 2]
```diff
@@ -759,6 +759,26 @@ object PositioningStrategies {
...
+        }
+    }
+
+    val RETURN_EXPRESSION: PositioningStrategy<KtReturnExpression> = object : PositioningStrategy<PsiElement>() {
+        override fun mark(element: PsiElement): List<TextRange> {
+            return if (element is KtReturnExpression) {
+                markElement(element.returnedExpression ?: element)
+            } else {
+                markElement(element)
+            }
```
Comment (KvanTTT) [inline compiler/frontend/src/org/jetbrains/kotlin/diagnostics/PositioningStrategies.kt:778]: Also, simplify:

```suggestion
            return markElement(if (element is KtReturnExpression) element.returnedExpression ?: element else element)
```

File: compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/expression/FirFunctionReturnTypeMismatch.kt [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+                val expectedArgs = functionReturnType.typeArguments.dropLast(1)
+                val actualArgs = returnExpressionType.typeArguments.dropLast(1)
+                return !compareTypesList(actualArgs, expectedArgs, context)
+            }
+        }
+
+        if (!AbstractTypeChecker.isSubtypeOf(context, returnExpressionType, functionReturnType))
+            return true
+
+        val returnResultType = returnExpression.result.typeRef.coneTypeSafe<ConeKotlinType>() ?: return false
```
Comment (mglukhikh) [inline compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/expression/FirFunctionReturnTypeMismatch.kt:58]: It seems to me this code repeats code above.

File: compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/expression/FirFunctionReturnTypeMismatch.kt [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+        if (!AbstractTypeChecker.isSubtypeOf(context, returnResultType, functionReturnType))
```
Comment (mglukhikh) [inline compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/expression/FirFunctionReturnTypeMismatch.kt:59]: And this check probably is redundant too.

File: compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/expression/FirFunctionReturnTypeMismatch.kt [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+    private fun isNeedToBeReported(
+        returnExpression: FirReturnExpression,
+        function: FirFunction<*>,
+        context: ConeInferenceContext
+    ): Boolean {
+        val functionReturnType = function.returnTypeRef.coneTypeSafe<ConeKotlinType>() ?: return false
+        val returnExpressionType = returnExpression.result.typeRef.coneTypeSafe<ConeKotlinType>() ?: return false
+        val functionLastTypeArg = functionReturnType.typeArguments.lastOrNull()?.type
+
+        if (functionReturnType.isFunctionalType(context.session) && functionLastTypeArg?.isUnit == true) {
```
Comment (mglukhikh) [inline compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/expression/FirFunctionReturnTypeMismatch.kt:44]: I'd try to extract this check or its part to some function and re-use it in FirInitializerTypeMismatchChecker.

**Comment by mglukhikh** [comment]
This PR is under consideration and testing now. At this moment I'd suggest it's "mostly Ok", but it's large and I can notice some problems later.

**Comment by mglukhikh** [comment]
Merged manually. Thank you. @vldF , you can look at my last commits if you are interested what additional fixes were done: https://github.com/JetBrains/kotlin/commits?author=mglukhikh.

**Pull request closed without merging**